File: src/com/krypton/updater/workers/DownloadWorker.java
/*
 * Copyright (C) 2021 AOSP-Krypton Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.krypton.updater.workers;

import static android.os.Process.THREAD_PRIORITY_BACKGROUND;
import static com.krypton.updater.util.Constants.UPDATE_PENDING;
import static com.krypton.updater.util.Constants.DOWNLOADING;
import static com.krypton.updater.util.Constants.FINISHED;
import static com.krypton.updater.util.Constants.FAILED;

import android.content.Context;
import android.os.FileUtils;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.work.Data;
import androidx.work.ForegroundInfo;
import androidx.work.ListenableWorker.Result;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.krypton.updater.model.data.BuildInfo;
import com.krypton.updater.model.data.DataStore;
import com.krypton.updater.model.data.DownloadStatus;
import com.krypton.updater.model.data.OTAFileManager;
import com.krypton.updater.R;
import com.krypton.updater.util.NotificationHelper;
import com.krypton.updater.util.Utils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;

public class DownloadWorker extends Worker {
    private static final String TAG = "DownloadWorker";
    private static final int UPDATE_DOWNLOAD_NOTIF_ID = 1002;
    private static final int BUF_SIZE = 8192; // 8 KB
    private final Context context;
    private final OTAFileManager ofm;
    private final Handler handler;
    private final HandlerThread thread;
    private final NotificationHelper helper;
    private final NotificationCompat.Builder notificationBuilder;
    private final DataStore dataStore;
    private int currPercent;
    private long currSize, totalSize;

    public DownloadWorker(Context context, WorkerParameters parameters,
            NotificationHelper helper, OTAFileManager ofm,
            DataStore dataStore) {
        super(context, parameters);
        this.context = context;
        this.helper = helper;
        this.ofm = ofm;
        this.dataStore = dataStore;
        thread = new HandlerThread(TAG, THREAD_PRIORITY_BACKGROUND);
        thread.start();
        handler = new Handler(thread.getLooper());
        notificationBuilder = helper.getDefaultBuilder()
            .setSmallIcon(android.R.drawable.stat_sys_download)
            .setNotificationSilent()
            .setOngoing(true)
            .setContentTitle(context.getString(R.string.downloading));
    }

    @Override
    public Result doWork() {
        final Data inputData = getInputData();
        totalSize = inputData.getLong(BuildInfo.FILE_SIZE, 0);
        int exitCode = download(inputData.getString(BuildInfo.URL),
            inputData.getString(BuildInfo.FILE_NAME), inputData.getString(BuildInfo.MD5));
        final Result result;
        switch (exitCode) {
            case -1:
                updateStatusAsync(FAILED);
                result = Result.failure();
                break;
            case 0:
                result = Result.retry();
                break;
            case 1:
                updateStatusAsync(FINISHED);
                result = Result.success();
                break;
            case 2:
                result = Result.success();
                break;
            default:
                result = Result.failure();
        }
        thread.quitSafely();
        return result;
    }

    /* return value:
     *   -1 for failure
     *    0 for retry
     *    1 for success
     *    2 if download was stopped (could be pause, cancel, or constraints not met)
     */
    private int download(String urlString, String fileName, String md5) {
        notificationBuilder.setContentText(fileName);
        setForegroundAsync(getForegroundInfo(0, true));
        final File file = new File(context.getExternalCacheDir(), fileName);
        long startByte = 0;
        boolean append = file.isFile();
        if (append) {
            currSize = file.length();
            final DownloadStatus downloadStatus = dataStore.getDownloadStatus();
            if (downloadStatus == null ||
                    downloadStatus.getDownloadedSize() != currSize) {
                currSize = 0;
                append = false;
            } else {
                startByte += currSize;
            }
        }
        URL url;
        try {
            url = new URL(urlString);
        } catch (MalformedURLException e) {
            Log.e(TAG, "Malformed url", e);
            helper.notifyOrToast(R.string.download_failed,
                R.string.invalid_url, handler);
            return -1;
        }
        HttpsURLConnection connection;
        try {
            connection = (HttpsURLConnection) url.openConnection();
        } catch (IOException e) {
            Log.d(TAG, "IOException when opening connection from url", e);
            return 0;
        }
        if (append) {
            connection.setRequestProperty("Range", String.format("bytes=%d-", startByte));
        }
        updateStatusAsync(DOWNLOADING);
        // Download starts here
        try (FileOutputStream outStream = new FileOutputStream(file, append);
                InputStream inStream = connection.getInputStream()) {
            final byte[] buffer = new byte[BUF_SIZE];
            int bytesRead = 0;
            while (!isStopped() && (bytesRead = inStream.read(buffer)) != -1) {
                outStream.write(buffer, 0, bytesRead);
                currSize += bytesRead;
                int tmp = (int) ((currSize*100)/totalSize);
                if (tmp > currPercent) {
                    currPercent = tmp;
                    setForegroundAsync(getForegroundInfo(currPercent, false)); // Update notification
                }
                updateProgressAsync(currSize, currPercent); // Update sharedPrefs
            }
            if (isStopped()) {
                return 2;
            }
        } catch (IOException e) {
            Log.d(TAG, "IOException when downloading content", e);
            return 0;
        }
        // Check if download is actually over
        if (currSize == totalSize) {
            if (Utils.computeMd5(file).equals(md5)) {
                if (copyFile(file, fileName)) {
                    file.delete();
                    helper.onlyNotify(R.string.download_finished, R.string.click_to_update);
                    // Mark as download finished and an update installation is pending
                    dataStore.setGlobalStatus(UPDATE_PENDING);
                    return 1;
                } else {
                    helper.notifyOrToast(R.string.download_failed,
                        R.string.copy_failed, handler);
                    return -1;
                }
            } else {
                helper.notifyOrToast(R.string.download_failed,
                    R.string.md5_check_failed, handler);
                return -1;
            }
        } else {
            return 0; // Retry download
        }
    }

    private void updateStatusAsync(int status) {
        handler.post(() -> dataStore.updateDownloadStatus(status));
    }

    private void updateProgressAsync(long size, int percent) {
        handler.post(() -> dataStore.updateDownloadProgress(size, percent));
    }

    private ForegroundInfo getForegroundInfo(int progress, boolean indeterminate) {
        return new ForegroundInfo(UPDATE_DOWNLOAD_NOTIF_ID, notificationBuilder
            .setProgress(100, progress, indeterminate)
            .build());
    }

    // Copy downloaded file to Downloads folder and then to ota dir (/data/kosp_ota)
    private boolean copyFile(File file, String fileName) {
        try (FileInputStream inStream = new FileInputStream(file)) {
            FileUtils.copy(file, Utils.getDownloadFile(fileName));
            return ofm.copyToOTAPackageDir(inStream);
        } catch (IOException e) {
            Log.e(TAG, "IOException when copying file " + file.getAbsolutePath(), e);
        }
        return false;
    }
}

File: app/src/main/java/com/krypton/updater/data/download/DownloadWorker.kt
/*
 * Copyright (C) 2021-2022 AOSP-Krypton Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.krypton.updater.data.download

import android.util.DataUnit
import android.util.Log

import com.krypton.updater.data.HashVerifier

import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.util.concurrent.TimeUnit

import javax.net.ssl.HttpsURLConnection

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull

/**
 * A worker whose job is to download the file from the given url.
 *
 * @property downloadFile the file to which the content should be downloaded to.
 * @property url the url to downloaded from.
 * @property fileSize the size of the file in bytes.
 * @property fileHash the SHA-512 hash of the file.
 */
class DownloadWorker(
    private val downloadFile: File,
    private val url: URL,
    private val fileSize: Long,
    private val fileHash: String,
) {
    private var downloadedBytes = 0L

    val channel = Channel<Long>(onBufferOverflow = BufferOverflow.DROP_OLDEST)

    /**
     * Run the worker.
     *
     * @return a [DownloadResult] indicating whether the work failed,
     *   or whether we should retry, or whether it was a success.
     */
    suspend fun run(): DownloadResult {
        logD("run")
        var resume = downloadFile.isFile
        if (resume) {
            downloadedBytes = downloadFile.length()
            logD("downloadedBytes = $downloadedBytes")
            if (downloadedBytes == fileSize) {
                logD("file already downloaded, verifying hash")
                if (HashVerifier.verifyHash(downloadFile, fileHash)) {
                    channel.send(downloadedBytes)
                    return DownloadResult.success()
                }
                Log.w(TAG, "File is corrupt, deleting")
                downloadFile.delete()
                downloadedBytes = 0
                resume = false
            }
        }
        logD("resume = $resume")
        val connectionResult = withContext(Dispatchers.IO) {
            openConnection(if (resume) "$downloadedBytes-" else null)
        }
        if (connectionResult.isFailure) {
            return DownloadResult.failure(connectionResult.exceptionOrNull())
        }
        val connection = connectionResult.getOrThrow()
        logD("connection opened")
        val downloadResult = withContext(Dispatchers.IO) {
            runCatching {
                FileOutputStream(downloadFile, resume).use { outStream ->
                    logD("output stream opened")
                    connection.inputStream.use { inStream ->
                        logD("input stream opened")
                        val buffer = ByteArray(DOWNLOAD_BUFFER_SIZE)
                        var bytesRead = inStream.read(buffer)
                        while (currentCoroutineContext().isActive && bytesRead > 0) {
                            outStream.write(buffer, 0, bytesRead)
                            downloadedBytes += bytesRead
                            channel.send(downloadedBytes)
                            bytesRead = inStream.read(buffer)
                        }
                        logD("isActive = ${currentCoroutineContext().isActive}")
                        outStream.flush()
                    }
                }
                logD("streams closed, download finished")
                connection.disconnect()
                logD("connection disconnected")
            }
        }
        channel.close()
        logD("channel closed")
        if (downloadResult.isFailure) {
            return DownloadResult.failure(downloadResult.exceptionOrNull())
        }
        return if (downloadedBytes == fileSize) {
            if (HashVerifier.verifyHash(downloadFile, fileHash))
                DownloadResult.success()
            else
                DownloadResult.failure(Throwable("SHA-512 hash doesn't match. Possible download corruption!"))
        } else {
            DownloadResult.retry()
        }
    }

    private suspend fun openConnection(range: String?): Result<HttpsURLConnection> {
        val connectionResult: Result<HttpsURLConnection>? =
            withTimeoutOrNull(CONNECTION_RETRY_TIMEOUT) {
                while (isActive) {
                    val openResult = runCatching {
                        url.openConnection() as HttpsURLConnection
                    }
                    if (openResult.isSuccess) {
                        val conn = openResult.getOrThrow()
                        if (range != null) {
                            conn.setRequestProperty("Range", "bytes=$range")
                        }
                        val responseCode = conn.responseCode
                        if (responseCode == HttpsURLConnection.HTTP_OK ||
                            responseCode == HttpsURLConnection.HTTP_CREATED ||
                            responseCode == HttpsURLConnection.HTTP_ACCEPTED ||
                            responseCode == HttpsURLConnection.HTTP_PARTIAL
                        ) {
                            Result.success(conn)
                        } else {
                            Log.e(TAG, "Connection failed with response code $responseCode")
                        }
                    }
                }
                Result.failure(Throwable("Current download job was cancelled"))
            }
        return connectionResult
            ?: Result.failure(Throwable("Timeout while establishing connection, retry"))
    }

    companion object {
        private const val TAG = "DownloadWorker"
        private val DEBUG: Boolean
            get() = Log.isLoggable(TAG, Log.DEBUG)

        private val DOWNLOAD_BUFFER_SIZE = DataUnit.KIBIBYTES.toBytes(256).toInt()

        private val CONNECTION_RETRY_TIMEOUT = TimeUnit.SECONDS.toMillis(5)

        private fun logD(msg: String) {
            if (DEBUG) Log.d(TAG, msg)
        }
    }
}